Web.Framework/HttpHandler.Factory.cs

```diff
@@ -194,147 +194,45 @@ public partial class HttpHandler
 
                 var lambda = Expression.Lambda<Func<HttpContext, RouteValueDictionary, RequestDelegate, Task>>(body, httpContextArg, routeValuesArg, nextArg);
 
-                var routeTemplate = method.RouteTemplate;
-                var matcher = routeTemplate == null ? null : new TemplateMatcher(routeTemplate, new RouteValueDictionary());
+                var routeTemplate = method.RoutePattern;
 
-                bindings.Add(new Binding
-                {
-#if DEBUG
-                    DebugExpression = lambda,
-#endif
-                    MethodInfo = method.MethodInfo,
-                    Invoke = lambda.Compile(),
-                    NeedForm = needForm,
-                    HttpMethod = httpMethod,
-                    Matcher = matcher
-                });
-            }
+                var invoker = lambda.Compile();
 
-            return next =>
-            {
-                return async context =>
-                {
-                    if (TryMatch(context, bindings, out var binding, out var routeValues))
+                var routeEndpointModel = new RouteEndpointModel(
+                    async httpContext =>
                     {
                         // Generating async code would just be insane so if the method needs the form populate it here
                         // so the within the method it's cached
-                        if (binding.NeedForm)
+                        if (needForm)
                         {
-                            await context.Request.ReadFormAsync();
+                            await httpContext.Request.ReadFormAsync();
                         }
 
-                        await binding.Invoke(context, routeValues, next);
-                        return;
-                    }
-
-                    await next(context);
-                };
-            };
-        }
-
-        private static bool TryMatch(HttpContext context, List<Binding> bindings, out Binding binding, out RouteValueDictionary routeValues)
-        {
-            // Scores
-            // nothing = 1
-            // method = 2
-            // route = 3
-            // method + route = 5
-            routeValues = null;
-            binding = null;
-
-            var matchValues = new RouteValueDictionary();
-            object match = null;
-            var matchScore = 0;
-
-            foreach (var b in bindings)
-            {
-                var score = 0;
+                        await invoker.Invoke(httpContext, httpContext.GetRouteData().Values, (c) => Task.CompletedTask);
```

Comment:
The invoker can probably be simplified since route values hang off of the request now.